pkg/controller/atlasproject/custom_roles.go

```diff
@@ -0,0 +1,201 @@
+				Name:      atlasAction.Action,
+				Resources: resources,
+			})
+		}
+
+		customRoles = append(customRoles, v1.CustomRole{
+			Actions:        actions,
+			InheritedRoles: inheritedRoles,
+			Name:           atlasCustomRole.RoleName,
+		})
+	}
+
+	return customRoles, nil
+}
+
+func deleteCustomRoles(ctx *workflow.Context, projectID string, currentCustomRoles []v1.CustomRole, desiredCustomRoles []v1.CustomRole) error {
+	toDelete := make([]v1.CustomRole, 0, len(currentCustomRoles))
+	desiredCustomRolesByName := mapCustomRolesByName(desiredCustomRoles)
+
+	for _, currentCustomRole := range currentCustomRoles {
+		if _, ok := desiredCustomRolesByName[currentCustomRole.Name]; !ok {
+			toDelete = append(toDelete, currentCustomRole)
+		}
+	}
+
+	ctx.Log.Debugw("Custom Roles to be deleted", "NumItems", len(toDelete))
+
+	for _, customRole := range toDelete {
+		_, err := ctx.Client.CustomDBRoles.Delete(context.Background(), projectID, customRole.Name)
+
+		if err != nil {
+			return err
+		}
+
+		ctx.Log.Debugw("Removed Custom Role from current AtlasProject", "custom role", customRole.Name)
+	}
+
+	return nil
+}
+
+func updateCustomRoles(ctx *workflow.Context, projectID string, currentCustomRoles []v1.CustomRole, desiredCustomRoles []v1.CustomRole) error {
+	toUpdate := make([]v1.CustomRole, 0, len(desiredCustomRoles))
+	currentCustomRolesByName := mapCustomRolesByName(currentCustomRoles)
+
+	for _, desiredCustomRole := range desiredCustomRoles {
+		customRole, ok := currentCustomRolesByName[desiredCustomRole.Name]
+
+		if !ok {
+			continue
+		}
+
+		if d := cmp.Diff(desiredCustomRole, customRole, cmpopts.EquateEmpty()); d != "" {
+			toUpdate = append(toUpdate, desiredCustomRole)
+		}
+	}
+
+	ctx.Log.Debugw("Custom Roles to be updated", "NumItems", len(toUpdate))
+
+	for _, customRole := range toUpdate {
+		data := customRole.ToAtlas()
+		// Patch fails when sending the role name in the body, needs clarification with cloud team
+		data.RoleName = ""
+		_, _, err := ctx.Client.CustomDBRoles.Update(context.Background(), projectID, customRole.Name, data)
+
+		if err != nil {
+			return err
+		}
+
+		ctx.Log.Debugw("Updated Custom Role in current AtlasProject", "custom role", customRole.Name)
+	}
+
+	return nil
+}
+
+func createCustomRoles(ctx *workflow.Context, projectID string, currentCustomRoles []v1.CustomRole, desiredCustomRoles []v1.CustomRole) error {
+	toCreate := make([]v1.CustomRole, 0, len(desiredCustomRoles))
+	currentCustomRolesByName := mapCustomRolesByName(currentCustomRoles)
+
+	for _, desiredCustomRole := range desiredCustomRoles {
+		if _, ok := currentCustomRolesByName[desiredCustomRole.Name]; !ok {
+			toCreate = append(toCreate, desiredCustomRole)
+		}
+	}
+
+	ctx.Log.Debugw("Custom Roles to be added", "NumItems", len(toCreate))
+
+	for _, customRole := range toCreate {
+		_, _, err := ctx.Client.CustomDBRoles.Create(context.Background(), projectID, customRole.ToAtlas())
+
+		if err != nil {
+			return err
```

Comment:
I'm not sure if it's worth returning an error here. For example, the user provided invalid data for the first role. And the Atlas API returns an error when it is created. Shouldn't we try to create all the other roles? Because the current code will try to create the first non-valid role and get stuck there.

Reply:
Other feature has structures in the spec and structures in the status. 
The status for each element contains related information: errors, IDs, and all related fields from the API response.
Thus, it is possible to correlate the element from the spec (in this case, it is a role) and the element in the status, thereby finding out the state of the element.

Reply:
Also, if the users provide two (or more) non-valid role specs, then they will receive only one error. So they will fix one role spec. And one more error will appear. 
They have to fix one error per edit. But the status for each role approach allows users to fix all errors.

Reply:
@Sugar-pack, it was an excellent point to add context to the errors and apply all the changes rather than fail quickly. I've applied those changes. However, for custom roles specifically, I believe is not necessary to have structured status once there's no complex operation going on on the main or in the nested structures.

Reply:
For two non-valid custom roles CR still contains only one error. 
```
status:
  authModes:
    - SCRAM
  conditions:
    - lastTransitionTime: '2022-11-02T06:16:13Z'
      status: 'False'
      type: Ready
    - lastTransitionTime: '2022-10-26T06:25:35Z'
      status: 'True'
      type: ValidationSucceeded
    - lastTransitionTime: '2022-11-01T12:52:17Z'
      status: 'True'
      type: ProjectReady
    - lastTransitionTime: '2022-11-02T06:16:20Z'
      message: >-
        POST
        https://cloud-qa.mongodb.com/api/atlas/v1.0/groups/635fd15382bd8930618b378c/customDBRoles/roles:
        400 (request "ATLAS_CUSTOM_ROLE_INVALID_NAME") The role name,
        "ShardingAdmin@@@@", is invalid. A custom role name must be a non-empty
        string made up of only ASCII letters, numbers, hyphens, and underscores,
        and should begin with a letter or number.
      reason: ProjectCustomRolesReady
      status: 'False'
      type: ProjectCustomRolesReady
  id: 635fd15382bd8930618b378c
  observedGeneration: 7
spec:
  customRoles:
    - actions:
        - name: LIST_SESSIONS
          resources:
            - cluster: true
        - name: KILL_ANY_SESSION
          resources:
            - cluster: true
      inheritedRoles:
        - database: admin
          name: enableSharding
        - database: admin
          name: backup
      name: ShardingAdmin@@@@
    - inheritedRoles:
        - database: test
          name: readWrite
        - database: test
          name: dbAdmin
      name: ShardingAdmin@@
  maintenanceWindow: {}
  name: Test project 1
  withDefaultAlertsSettings: true

```

Reply:
The status for each role isn't just about complex entities. If any of the roles is invalid, then the CR clearly shows which roles are created and which are not. For example:
CR spec contains roles:
```
role1 - non-valid name
role2 - non-valid action
role3 - valid
```
The current approach:
`role1 contains a non-valid name, and role2 contains a non-valid action`
Approach with status for each role:
```
role1 - failed: contains a non-valid name
role2 - failed: contains a non-valid action
role3 - created
```
The user is explicitly told the status of each role. And user sees that role3  has been created. But in the current approach, the user needs to know additional information about our operator. Moreover, all other features work through individual statuses for each element.

Reply:
@Sugar-pack While I agree the status for each role is easier to read, I believe that the API and the entity(custom role) itself don't provide enough information to make a clear set of infos which represents the current status of a certain custom role.
what would be the status when updating it? what if we update it again and again? and what if we delete it?

maybe @igor-karpukhin and @fabritsius have an opinion here?